lib/CL/pocl_llvm_wg.cc

```diff
@@ -160,39 +155,227 @@ static TargetMachine *GetTargetMachine(cl_device_id device, Triple &triple) {
+
+public:
+  TwoStagePoclModulePassManager() = default;
+  llvm::Error build(cl_device_id Dev, const std::string &S1_Pipeline,
+                    unsigned S1_OLevel, unsigned S1_SLevel,
+                    const std::string &S2_Pipeline, unsigned S2_OLevel,
+                    unsigned S2_SLevel);
+  void run(llvm::Module &Bitcode);
+};
+
+llvm::Error TwoStagePoclModulePassManager::build(
+    cl_device_id Dev, const std::string &S1_Pipeline, unsigned S1_OLevel,
+    unsigned S1_SLevel, const std::string &S2_Pipeline, unsigned S2_OLevel,
+    unsigned S2_SLevel) {
+  llvm::Error E1 = Stage1.build(S1_Pipeline, S1_OLevel, S1_SLevel, Dev);
+  if (E1)
+    return E1;
+
+  return Stage2.build(S2_Pipeline, S2_OLevel, S2_SLevel, Dev);
+}
+
+void TwoStagePoclModulePassManager::run(llvm::Module &Bitcode) {
+  Stage1.run(Bitcode);
+  Stage2.run(Bitcode);
+}
+
+static std::map<cl_device_id, TwoStagePoclModulePassManager *> kernelPasses;
+#endif
+
+void clearKernelPasses() {
+  for (auto i = kernelPasses.begin(), e = kernelPasses.end(); i != e; ++i) {
+    delete i->second;
+  }
+
+  kernelPasses.clear();
+}
+
+enum class PassType {
+  Module,
+  CGSCC,
+  Function,
+  Loop,
+};
+
+// for legacy PM, add each Pass to the pass pipeline;
+// for new PM, additionally they must be registered in registerPassBuilderPasses
+static void addPass(std::vector<std::string> &Passes, std::string PassName,
+                    PassType T = PassType::Function) {
+#if LLVM_MAJOR < MIN_LLVM_NEW_PASSMANAGER
+  Passes.push_back(PassName);
+#else
+  std::string Temp;
+  switch (T) {
+  case PassType::Module:
+    Passes.push_back(PassName);
+    break;
+  case PassType::CGSCC:
+    Temp = "cgscc(" + PassName + ")";
+    Passes.push_back(Temp);
+    break;
+  case PassType::Function:
+    Temp = "function(" + PassName + ")";
+    Passes.push_back(Temp);
+    break;
+  case PassType::Loop:
+    Temp = "function(loop(" + PassName + "))";
+    Passes.push_back(Temp);
+    break;
+  default:
+    POCL_ABORT("unknown pass type");
+  }
+#endif
+}
+
+// for legacy PM, add each Analysis to the pass pipeline;
+// for new PM, add them with proper nesting...  X(Y(Require<>))
+static void addAnalysis(std::vector<std::string> &Passes, std::string PassName,
+                        PassType T = PassType::Function) {
+#if LLVM_MAJOR < MIN_LLVM_NEW_PASSMANAGER
+  Passes.push_back(PassName);
+#else
+  std::string Temp;
+  PassName = "require<" + PassName + ">";
+  switch (T) {
+  case PassType::Module:
+    Passes.push_back(PassName);
+    break;
+  case PassType::CGSCC:
+    Temp = "cgscc(" + PassName + ")";
+    Passes.push_back(Temp);
+    break;
+  case PassType::Function:
+    Temp = "function(" + PassName + ")";
+    Passes.push_back(Temp);
+    break;
+  case PassType::Loop:
+    Temp = "function(loop(" + PassName + "))";
+    Passes.push_back(Temp);
+    break;
+  default:
+    POCL_ABORT("unknown pass type");
+  }
+#endif
+}
+
+static void addStage1PassesToPipeline(cl_device_id Dev,
```

Comment:
Again this calls for minimal documentation what are stage 1 and stage 2 and why, etc.